File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/subscription/ModeBRequestScreen.kt
package lk.mageride.passenger.subscription

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.home.VehicleLabels
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.LabelledTextField
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.MageRideTextLink
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.AccessRequestStatus
import lk.mageride.shared.data.models.ServiceMode

/**
 * SCR-PA-024 — "Private transport", the Mode B access request.
 *
 * The wireframe: a `MODE B` badge card explaining that *"the owner approves in the Driver App"*, a
 * **Vehicle ID** field, the Pending/Accepted/Rejected chip, and a **Send request** CTA pinned to
 * the bottom.
 *
 * **Two doors, one screen** (AL-23). A Mode B marker on SCR-PA-010 hands over the vehicle id it
 * just drew; the drawer's *"Private transport"* row opens the same screen with an empty field. The
 * route argument is optional for exactly that reason — see `PassengerRoute.ModeBRequest`.
 *
 * **A Mode B marker never opens SCR-PA-007.** That popup is Mode A's; a private vehicle a
 * passenger has no grant for has nothing to show them but this.
 */
@Composable
internal fun ModeBRequestScreen(onBack: () -> Unit, onOpenSubscriptions: () -> Unit, model: ModeBRequestViewModel) {
    val state by model.state.collectAsStateWithLifecycle()

    Column(modifier = Modifier.fillMaxSize()) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = onBack) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                text = stringResource(R.string.mode_b_request_title),
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            ExplainerCard()

            LabelledTextField(
                label = stringResource(R.string.mode_b_vehicle_id),
                value = state.vehicleId,
                onValueChange = model::onVehicleIdChange,
                placeholder = VEHICLE_ID_EXAMPLE,
                enabled = !state.sending && state.existing == null,
                imeAction = ImeAction.Done,
            )

            Text(
                text = stringResource(
                    if (state.prefilled) R.string.mode_b_from_marker else R.string.mode_b_type_id,
                ),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            state.status?.let { DecisionCard(status = it, onOpenSubscriptions = onOpenSubscriptions) }

            state.error?.let { InlineError(message = stringResource(it)) }

            // The wireframe's `spacer` — the CTA is pinned to the bottom of the body (US-24.1's
            // rule for SCR-PA-002, applied here because the wireframe draws the same layout).
            Spacer(modifier = Modifier.weight(1f))

            MageRideCta(
                label = stringResource(R.string.mode_b_send_request),
                onClick = model::send,
                enabled = state.canSend,
                loading = state.sending,
            )
        }
    }
}

/** The wireframe's filled `MODE B` card. */
@Composable
private fun ExplainerCard() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                MaterialTheme.colorScheme.surfaceVariant,
                RoundedCornerShape(MageRideTheme.radius.md),
            )
            .padding(MageRideTheme.spacing.sm),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
    ) {
        Text(
            // C078's own mode copy, not a second key for the same words — D2' §0.2 has one Mode B
            // label and `StringResourceTest` reads two identical values as a translation nobody did.
            text = stringResource(VehicleLabels.modeName(ServiceMode.B)),
            modifier = Modifier
                .background(MageRideTheme.mode.modeB, RoundedCornerShape(MageRideTheme.radius.sm))
                .padding(horizontal = MageRideTheme.spacing.xs, vertical = MageRideTheme.spacing.xxs),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onPrimary,
        )
        Text(
            text = stringResource(R.string.mode_b_explainer),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

/**
 * US-4.6's chip, as the wireframe's tinted card.
 *
 * **Pending is the state this screen can actually observe.** Accepted is inferred from a
 * subscription existing (see [ModeBRequestViewModel]); Rejected has no signal on this surface at
 * all and is drawn only because the enum can carry it — a rejection reaches a passenger through
 * the owner, not through the platform. Recorded in the C082 handoff.
 */
@Composable
private fun DecisionCard(status: AccessRequestStatus, onOpenSubscriptions: () -> Unit) {
    val tint = when (status) {
        AccessRequestStatus.PENDING -> MageRideTheme.status.warning
        AccessRequestStatus.ACCEPTED -> MageRideTheme.status.success
        AccessRequestStatus.REJECTED -> MaterialTheme.colorScheme.error
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(tint.copy(alpha = CARD_TINT), RoundedCornerShape(MageRideTheme.radius.md))
            .padding(MageRideTheme.spacing.sm),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
    ) {
        Text(
            text = stringResource(
                when (status) {
                    AccessRequestStatus.PENDING -> R.string.mode_b_status_pending
                    AccessRequestStatus.ACCEPTED -> R.string.mode_b_status_accepted
                    AccessRequestStatus.REJECTED -> R.string.mode_b_status_rejected
                },
            ),
            style = MaterialTheme.typography.titleSmall,
            color = MaterialTheme.colorScheme.onSurface,
        )
        Text(
            text = stringResource(
                when (status) {
                    // US-4.7's first month free, which is the fleet's offer and not the platform's.
                    AccessRequestStatus.PENDING -> R.string.mode_b_pending_note

                    AccessRequestStatus.ACCEPTED -> R.string.mode_b_accepted_note

                    AccessRequestStatus.REJECTED -> R.string.mode_b_rejected_note
                },
            ),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        if (status == AccessRequestStatus.ACCEPTED) {
            MageRideTextLink(
                label = stringResource(R.string.mode_b_open_subscriptions),
                onClick = onOpenSubscriptions,
            )
        }
    }
}

/** The wireframe's status cards are a wash of their own colour, not a solid fill. */
private const val CARD_TINT = 0.16f

/**
 * The wireframe's example Vehicle ID, as the field's placeholder.
 *
 * A registration handle is the same characters in all three scripts — the same argument
 * `LanguageNames` and `PhoneNumber` make for the language endonyms and the `+94` prefix — so it is
 * a Kotlin constant rather than three identical `strings.xml` values, which `StringResourceTest`
 * reads as a translation nobody did.
 */
private const val VEHICLE_ID_EXAMPLE = "MR-VEH-48213"

File: apps/passenger-ios/PassengerApp/Subscription/ModeBRequestScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-024 — *"Private transport"*, the Mode B access request.
///
/// The cell: `‹ Back · Private transport`, a `card fill` carrying a `MODE B` badge and the line
/// *"Request access to track a private vehicle. Owner approves in the Driver App."*, a **Vehicle ID**
/// field, the Pending/Accepted/Rejected card, a `spacer`, and a **Send request** CTA pinned to the
/// bottom.
///
/// **Two doors, one screen** (AL-23). A Mode B marker on SCR-PI-010 hands over the vehicle id it just
/// drew (``LiveMapModel/onMarkerTapped(_:)`` routes it there); the Menu tab's *"Private transport"*
/// row opens the same screen with an empty field. The route's associated value is optional for exactly
/// that reason — see ``PassengerRoute/modeBRequest(vehicleId:)``.
///
/// **A Mode B marker never opens SCR-PI-007.** That popup is Mode A's; a private vehicle a passenger
/// has no grant for has nothing to show them but this.
@MainActor
struct ModeBRequestScreen: View {

    @StateObject private var model: ModeBRequestModel

    /// SCR-PI-025, from the accepted card's link.
    let onOpenSubscriptions: () -> Void

    init(
        vehicleId: String?,
        subscriptions: SubscriptionRepository,
        sessions: PassengerSessions,
        keys: IdempotencyKeys,
        onOpenSubscriptions: @escaping () -> Void
    ) {
        _model = StateObject(
            wrappedValue: ModeBRequestModel(
                vehicleId: vehicleId,
                subscriptions: subscriptions,
                sessions: sessions,
                keys: keys
            )
        )
        self.onOpenSubscriptions = onOpenSubscriptions
    }

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            explainer

            LabelledTextField(
                labelKey: "mode_b_vehicle_id",
                value: vehicleIdBinding,
                placeholder: ModeBRequestScreen.vehicleIdExample,
                supportingKey: model.state.isPrefilled ? "mode_b_from_marker" : "mode_b_type_id",
                autocapitalisation: .characters
            )
            .disabled(model.state.isSending || model.state.existing != nil)

            if let status = model.state.status {
                decision(status)
            }

            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
            }

            // The wireframe's `spacer` — the CTA is pinned to the bottom of the body.
            Spacer(minLength: MageRideSpacing.md)

            Button {
                Task { await model.send() }
            } label: {
                Text(key: "mode_b_send_request")
            }
            .buttonStyle(.mageCta(loading: model.state.isSending))
            .disabled(!model.state.canSend)
        }
        .padding(MageRideSpacing.md)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "mode_b_request_title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { await model.loadExisting() }
    }

    // MARK: -

    /// The wireframe's filled `MODE B` card.
    private var explainer: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
            // C096's own mode copy, not a second key for the same words — D2' §0.2 has one Mode B
            // label and `LocalizationTests` reads two identical values as a translation nobody did.
            SolidBadge(text: ModeToken.b.nameKey.localised, color: ModeToken.b.color)

            Text(key: "mode_b_explainer")
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(MageRideSpacing.sm)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }

    /// US-4.6's chip, as the wireframe's tinted card.
    ///
    /// **Pending is the state this screen can actually observe.** Accepted is inferred from a
    /// subscription existing (see ``ModeBRequestModel``); Rejected has no signal on this surface at
    /// all and is drawn only because the enum can carry it — a rejection reaches a passenger through
    /// the owner, not through the platform. Restated in the C100 handoff.
    private func decision(_ status: AccessRequestStatus) -> some View {
        let labels = ModeBRequestScreen.decisionLabels(for: status)

        return VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
            StatusPill(titleKey: labels.titleKey, tone: labels.tone)

            Text(key: labels.noteKey)
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .fixedSize(horizontal: false, vertical: true)

            if model.state.isAccepted {
                TextLink(key: "mode_b_open_subscriptions") { onOpenSubscriptions() }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(MageRideSpacing.sm)
        .background(
            // The wireframe's status cards are a wash of their own colour, not a solid fill — the
            // same treatment ``StatusPill`` gives its own background.
            ModeBRequestScreen.tint(for: labels.tone).opacity(0.14),
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }

    private var vehicleIdBinding: Binding<String> {
        Binding(get: { model.state.vehicleId }, set: model.onVehicleIdChange)
    }

    /// The three decisions, as copy and a tone.
    ///
    /// A `static` function rather than a computed property on ``AccessRequestStatus`` for
    /// ``RideStateLabel``'s reason: the mapping from a wire enum onto trilingual copy belongs beside
    /// the screen that draws it, and there is exactly one of those.
    ///
    /// `nonisolated` because it is a pure table and this type is `@MainActor` — a suite asserting the
    /// mapping should not have to be main-actor isolated to read it.
    nonisolated static func decisionLabels(for status: AccessRequestStatus) -> (titleKey: String,
                                                                               noteKey: String,
                                                                               tone: StatusPill.Tone) {
        if status == AccessRequestStatus.accepted {
            return ("mode_b_status_accepted", "mode_b_accepted_note", .ok)
        }
        if status == AccessRequestStatus.rejected {
            return ("mode_b_status_rejected", "mode_b_rejected_note", .error)
        }
        // US-4.7's first month free, which is the fleet's offer and not the platform's.
        return ("mode_b_status_pending", "mode_b_pending_note", .warning)
    }

    private static func tint(for tone: StatusPill.Tone) -> Color {
        switch tone {
        case .ok: return MageRideColor.success
        case .warning: return MageRideColor.warning
        case .error: return MageRideColor.error
        case .info: return MageRideColor.secondary
        case .muted: return MageRideColor.onSurfaceVariant
        }
    }

    /// The wireframe's example Vehicle ID, as the field's placeholder.
    ///
    /// A registration handle is the same characters in all three scripts — the same argument
    /// ``LanguageDisplay``, ``PhoneNumber`` and ``MoneyFormat/prefix`` make — so it is a Swift constant
    /// rather than three identical `.strings` values, which `LocalizationTests` reads as a translation
    /// nobody did.
    static let vehicleIdExample = "MR-VEH-48213"
}
